server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -604,32 +608,69 @@ public ClusterState addIndexTemplateV2(
             return currentState;
         }
 
-        validate(name, finalIndexTemplate);
-        validateDataStreamsStillReferenced(currentState, name, finalIndexTemplate);
+        validateIndexTemplateV2(name, finalIndexTemplate, currentState);
+        logger.info(
+            "{} index template [{}] for index patterns {}",
+            existing == null ? "adding" : "updating",
+            name,
+            template.indexPatterns()
+        );
+        return ClusterState.builder(currentState).metadata(Metadata.builder(currentState.metadata()).put(name, finalIndexTemplate)).build();
+    }
+
+    private void validateIndexTemplateV2(String name, ComposableIndexTemplate indexTemplate, ClusterState currentState) {
```

Comment:
@dakrone I'm curious what your thoughts are around this particular change to `MetadataIndexTemplateService`.

The `MetadataCreateDataStreamService` now adds the start_time and end_time index settings, but when this class validates a composable index template then without this change this would fail (index.mode setting validation requires the start_time and end_time index settings to be present if index.mode=time_series). This change adds the start_time and end_time index settings only when validating the template. These settings don't get added the the template instance that get stored. This way composable index templates for data streams don't need to specify the start_time and end_time settings, which would be weird, because it is a template and not an actual index (which means dummy values would be defined). 

One design decision is that a tsdb index can also be created without data streams and therefor when a regular index with index.mode=time_series is created then validation of start_time and end_time index settings should always kick in.

Reply:
I think doing the validation this way is okay, I understand why we need to add it for validation, even though it's calculated at index creation time.

I do have a question about applying the setting though, could we use the `IndexSettingProvider` for adding it similar to the way that we add the setting for data tiers? We could potentially pass `Metadata` in to the method so that it could tell whether this was the initial index for a data stream. That seems like it would be a little cleaner than injecting it in `MetadataCreateDataStreamService`.

What do you think?

Reply:
Oh also, I *think* we'll need to add this mocking to the temporary index creation in the `TransportSimulateIndexAction` and `TransportSimulateIndexTemplateAction`, because I think they might throw an error because these settings are missing if someone had `index.mode: time_series` in their template.

Reply:
@dakrone what do you think about this approach that uses `IndexSettingProvider `: 
7152640e66ff6278287c7eefe3649f766164c5b0

Reply:
Thanks, I think this looks cleaner to me, and re-uses our existing infrastructure.